Report tool now check free disk space before starting

klaren opened the pull request:
Reports can be quite large if both transaction files and heap dump is included.

This PR makes sure that the available disk space is large enough before starting to generate a report to avoid running out of disk space. The estimation of the final size is a pure guess and pessimistically chosen.

It also introduces a `--force` argument to the tool that can be used to ignore the warning about the free disk space.

klaren committed:
Report tool now check free disk space before starting

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/dbms/diagnostics/jmx/JmxDump.java:
```diff
@@ -181,26 +181,44 @@ public String destinationPath()
+                }
+                catch ( IOException e )
+                {
+                    progress.error( "Unable to determine size of neo4j instance memory", e );
```
note: throw exception instead

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java:
```diff
@@ -75,7 +75,23 @@ public void dump( Set<String> classifiers, Path destination, DiagnosticsReporter
+            String message =
+                    String.format( "WARNING: Free available disk space for %s is %s, worst case estimate is %s",
+                            destination.getFileName(), Format.bytes( freeSpace ), Format.bytes( estimatedFinalSize ) );
+            if ( !progress.shouldIgnorePotentialFullDisk( message ) )
```
let's change this condition from `!shouldIgnore` to `continueOnFull` to remove negation

klaren replied:
Still a negation though :sweat_smile: 
Choosing `stopOnPotentialFullDisk`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporter.java:
```diff
@@ -75,7 +75,23 @@ public void dump( Set<String> classifiers, Path destination, DiagnosticsReporter
+        Files.createDirectories( destinationFolder );
+
+        // Estimate an upper bound of the final size and make sure it will fit
+        long estimatedFinalSize = sources.stream().mapToLong(
```
lets extract this space check into separate helper method instead

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/diagnostics/DiagnosticsReporterProgressInteractions.java:
```diff
@@ -48,6 +48,14 @@
+     * @param message with information about the disk space.
+     * @return true if the execution should continue even if risking to fill up the disk.
+     */
+    boolean shouldIgnorePotentialFullDisk( String message );
```
it feels a bit wrong that its a progress reporter who decide should or shouldn't we ignore potential OOD problems.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/diagnostics/InteractiveProgress.java:
```diff
@@ -102,6 +104,33 @@ public void error( String msg, Throwable throwable )
+        while ( true )
+        {
+            out.print( "Ignore available disk space warning? [yN]: " );
+            String answer = System.console().readLine();
```
`answer` can be null just use `StringUtils.isEmpty` instead

MishaDemianenko reviewed (changes requested):
Nice in general, couple of comments here and there

klaren committed:
Simplified full disk check to just end execution if not --force is passed
If heap dump size cannot be determined, the execution not ends with an exception.

klaren commented:
@MishaDemianenko Addressed comments, please have another look :)
